src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityDistillationTower.java

```diff
@@ -33,54 +29,58 @@
 import net.minecraft.util.text.ITextComponent;
 import net.minecraft.util.text.TextFormatting;
 import net.minecraftforge.fluids.FluidStack;
-import net.minecraftforge.fluids.IFluidTank;
-import net.minecraftforge.fluids.capability.IFluidHandler;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 import net.minecraftforge.items.IItemHandlerModifiable;
 
-import com.cleanroommc.modularui.utils.FluidTankHandler;
-import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import org.jetbrains.annotations.NotNull;
 
 import java.util.List;
 import java.util.function.Function;
-import java.util.stream.Collectors;
 
 import static gregtech.api.util.RelativeDirection.*;
 
-public class MetaTileEntityDistillationTower extends RecipeMapMultiblockController {
+public class MetaTileEntityDistillationTower extends RecipeMapMultiblockController
+                                             implements DistillationTowerLogicHandler.IDistillationTower {
 
-    private final boolean useAdvHatchLogic;
-
-    protected int layerCount;
-    protected List<IFluidHandler> orderedFluidOutputs;
+    protected DistillationTowerLogicHandler handler;
 
+    @SuppressWarnings("unused") // backwards compatibility
     public MetaTileEntityDistillationTower(ResourceLocation metaTileEntityId) {
         this(metaTileEntityId, false);
     }
 
     public MetaTileEntityDistillationTower(ResourceLocation metaTileEntityId, boolean useAdvHatchLogic) {
         super(metaTileEntityId, RecipeMaps.DISTILLATION_RECIPES);
-        this.useAdvHatchLogic = useAdvHatchLogic;
         if (useAdvHatchLogic) {
-            this.recipeMapWorkable = new DistillationTowerRecipeLogic(this);
-        }
+            DistillationTowerRecipeLogic logic = new DistillationTowerRecipeLogic(this);
+            this.recipeMapWorkable = logic;
+            this.handler = new DistillationTowerLogicHandler(this, logic);
+        } else this.handler = null;
```

Comment:
Is there any reason to even make this logic optional? I can't really think of a good case for when you wouldn't want it.

Reply:
The logic will explode if an addon has a child class of the DT that has a different structure and the two information-gathering methods are not overriden. The optionality is just so that in those cases their child class still works properly.